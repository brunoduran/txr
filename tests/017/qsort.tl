(load "../common")

(with-dyn-lib (libc)
  ;; FFI for sorting char *array[] with qsort.
  (deffi qsort "qsort" void ((ptr (array str)) size-t size-t closure))
  (deffi-cb qsort-cb int ((ptr str-d) (ptr str-d)))

  ;; FFI for sorting wchar_t *array[] with qsort.
  (deffi qsortw "qsort" void ((ptr (array wstr)) size-t size-t closure))
  (deffi-cb qsortw-cb int ((ptr wstr-d) (ptr wstr-d))))

;; sort vector of strings as char *
(let ((vec #("the" "quick" "brown" "fox" "jumped" "over" "the" "lazy" "dogs")))
  (prinl vec)
  (qsort vec (length vec) (sizeof str) [qsort-cb cmp-str])
  (prinl vec))

;; sort vector of strings as wchar_t *
(let ((vec #("the" "quick" "brown" "fox" "jumped" "over" "the" "lazy" "dogs")))
  (prinl vec)
  (qsortw vec (length vec) (sizeof str) [qsortw-cb cmp-str])
  (prinl vec))

;; abort callback with non-local transfer
(let ((vec #("the" "quick" "brown" "fox" "jumped" "over" "the" "lazy" "dogs")))
  (prinl vec)
  (prinl (block foo
           (qsort vec (length vec) (sizeof str)
                  [qsort-cb (lambda (x y) (return-from foo 42))])))
  (prinl vec))

;; abort callback with non-local transfer
(let ((vec #("the" "quick")))
  (prinl vec)
  (prinl (block foo
           (qsortw vec (length vec) (sizeof str)
                   [qsortw-cb (lambda (x y) (return-from foo 42))])))
  (prinl vec))
