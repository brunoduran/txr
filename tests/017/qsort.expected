#("the" "quick" "brown" "fox" "jumped" "over" "the" "lazy" "dogs")
#("brown" "dogs" "fox" "jumped" "lazy" "over" "quick" "the" "the")
#("the" "quick" "brown" "fox" "jumped" "over" "the" "lazy" "dogs")
#("brown" "dogs" "fox" "jumped" "lazy" "over" "quick" "the" "the")
#("the" "quick" "brown" "fox" "jumped" "over" "the" "lazy" "dogs")
42
#(nil nil nil nil nil nil nil nil nil)
#("the" "quick")
42
#(nil nil)
